com.unity.multiplayer.mlapi/Runtime/AOI/ClientObjectMap.cs

```diff
@@ -0,0 +1,110 @@
+
+        // to dynamically compute objects to be added each 'QueryFor' call,
+        //  assign this delegate to your handler
+        public delegate void QueryDelegate(CLIENT c, HashSet<OBJECT> results);
+        public QueryDelegate OnQuery;
+
+        public delegate void BypassDelegate(HashSet<OBJECT> results);
+        public BypassDelegate OnBypass;
+
+        public ClientObjMapNode() : base()
+        {
+            ChildNodes = new List<ClientObjMapNode<CLIENT, OBJECT>>();
+        }
+
+        // externally-called object query function.  Call this on your root
+        //  ClientObjectMapNode.  The passed-in hash set will contain the results.
+        public void QueryFor(CLIENT client, HashSet<OBJECT> results)
+        {
+            if (Bypass)
+            {
+                OnBypass(results);
+            }
+            else
+            {
+                if (OnQuery != null)
+                {
+                    OnQuery(client, results);
+                }
+
+                foreach (var c in ChildNodes)
+                {
+                    c.QueryFor(client, results);
+                }
+            }
+        }
+
+        // Called when a given object is about to be despawned.  The OnDespawn
+        //  delegate gives each node a chance to do its own handling (e.g. removing
+        //  the object from a cache)
+        public void DespawnCleanup(OBJECT o)
+        {
+            if (OnDespawn != null)
+            {
+                OnDespawn(o);
+            }
+
+            foreach (var c in ChildNodes)
+            {
+                c.DespawnCleanup(o);
+            }
+        }
+
+        // Add a new child node.  Currently, there is no way to remove a node
+        public void AddNode(ClientObjMapNode<CLIENT, OBJECT> @new)
+        {
+            ChildNodes.Add(@new);
+        }
+
+        private List<ClientObjMapNode<CLIENT, OBJECT>> ChildNodes;
+        public bool Bypass = false;
+   }
+
+   // Static node type.  Objects can be added / removed as desired.
+   //  When the Query is done, these objects are grafted in without
+   //  any per-object computation.
+   public class ClientObjMapNodeStatic<CLIENT, OBJECT> : ClientObjMapNode<CLIENT, OBJECT>
+    {
+        public ClientObjMapNodeStatic() : base()
+        {
+            alwaysRelevant = new HashSet<OBJECT>();
+
+            // when we are told an object is despawning, remove it from our list
+            OnDespawn = Remove;
+
+            // for our query, we simply union our static objects with the results
+            //  more sophisticated methods might be explored later, like having the results
+            //  list contain not ust
+            OnQuery = delegate(CLIENT client, HashSet<OBJECT> results)
+            {
+                results.UnionWith(alwaysRelevant);
+            };
+        }
+
+        // Add a new item to our static list
+        public void Add(OBJECT o) // ref?
+        {
+            alwaysRelevant.Add(o);
+        }
```

Comment:
when and how we're adding new objects? I'd like to ask for simple example usages of both non-static (base one) and static node implementation. maybe that could be a separate PR but I'm just curious.